File: book/packt/Scala.Functional.Programming.Patterns/Chapter 3/code/scala/Parser.scala
package chapter03

import scala.annotation.tailrec

object Parser extends App {

	val Number = """^(\d+).*""".r      // 1
		val LParen = """^[(].*""".r

		def factor(f: String): (String, Int) = f match {
			case Number(d) => (f.drop(d.length), d.toInt) // 2
				case LParen(_*) => {
					val p = expr(f.drop(1), 0)  // 3
						val e = p._1
						if (e.take(1) == ")") {  // 4
							(e.drop(1), p._2)
						} else {
							throw new IllegalArgumentException("Right bracket missing")
						}
				}
			case _ => throw new IllegalArgumentException("Number or sub-expression expected")
		}

	@tailrec
		def term(t: String, acc: Int): (String, Int) = {
			val p = factor(t)
				val e = p._1

				if (e.take(1) == "*") {                 // 5
					term(e.drop(1), acc * p._2)       // 6
				} else {
					(e, acc * p._2)                           // 7
				}
		}

	@tailrec
		def expr(s: String, acc: Int): (String, Int) = {
			val p = term(s, 1)
				val e = p._1

				if (e.take(1) == "+") {                  // 8
					expr(e.drop(1), acc + p._2)         // 9
				} else {
					(e, acc + p._2)                             // 10
				}
		}

	def expr(s: String): Int = {
		val e = expr(s, 0)
			e._2
	}

	val p = expr("(1+2)*3*(2+4)")
	println(p)
}

File: book/packt/Scala.Functional.Programming.Patterns/Chapter 3/code/java/Parser.java
package chapter03;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.lang3.Validate;

public class Parser {

	private static final String L_BRACKET = "(";
	private static final String R_BRACKET = ")";

	private class Tokenizer {

		private static final String NUM_PATTERN = "(\\d+).*";
		private String s;
		private final Pattern p;

		public Tokenizer(final String s) {
			this.s = s;
			this.p = Pattern.compile(NUM_PATTERN);
		}

		public boolean nextTokenIs(final String tok) {
			if (s.isEmpty()) {
				return false;
			}
			return s.startsWith(tok);
		}

		public void consume(final String tok) {
			s = s.replaceFirst(Pattern.quote(tok), "");
		}

		public boolean nextTokenIsNumber() {
			return s.matches(NUM_PATTERN);
		}

		public int consumeANumber() {
			if (s.matches(NUM_PATTERN)) {
				final Matcher m = p.matcher(s);
				Validate.isTrue(m.matches(), "Could not extract number from <"
						+ s + ">");
				final String numStr = m.group(1);
				s = s.replaceFirst(Pattern.quote(numStr), ""); 
				return Integer.valueOf(numStr);
			}

			throw new IllegalArgumentException("Number expected");
		}

	};

	private final Tokenizer tokenizer;

	public Parser(final String s) {
		tokenizer = new Tokenizer(s);
	}

	private int factor() {
		if (tokenizer.nextTokenIsNumber()) {
			final int num = tokenizer.consumeANumber(); // 1
			return num;
		}
		if (tokenizer.nextTokenIs(L_BRACKET)) { 
			tokenizer.consume(L_BRACKET); // 2
			final int num = expr();
			if (!tokenizer.nextTokenIs(R_BRACKET)) {
				throw new IllegalArgumentException("Syntax error - ) missing");
			}
			tokenizer.consume(R_BRACKET);  // 3
			return num;
		}
		throw new IllegalArgumentException("Either number or ( expected");
	}

	private int term() {
		int val = factor();     // 4
		while (tokenizer.nextTokenIs("*")) { // 5
			tokenizer.consume("*");
			val *= factor(); //  6
		}
		return val;
	}


	public int expr() {
		int val = term();           //  7
		while (tokenizer.nextTokenIs("+")) {   // 8      
			tokenizer.consume("+");
			val += term();   // 9
		}
		return val;
	}

	

}
